backstory/src/main/java/com/google/sps/vision/VisionManagerImpl.java

```diff
@@ -0,0 +1,181 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.vision;
+
+import com.google.cloud.vision.v1.AnnotateImageRequest;
+import com.google.cloud.vision.v1.AnnotateImageResponse;
+import com.google.cloud.vision.v1.BatchAnnotateImagesResponse;
+import com.google.cloud.vision.v1.EntityAnnotation;
+import com.google.cloud.vision.v1.Feature;
+import com.google.cloud.vision.v1.Image;
+import com.google.cloud.vision.v1.ImageAnnotatorClient;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.protobuf.ByteString;
+import java.io.IOException;
+import java.lang.IllegalArgumentException;
+import java.util.ArrayList;
+import java.util.List;
+
+public final class VisionManagerImpl implements VisionManager {
+  private final byte[] rawImageData;
+  private final List<EntityAnnotation> labelAnnotations;
+
+  /**
+   * VisionManager Manages the gathering and packaging of Vision API image analytics.
+   * A VisionManager object represents an image annotated with analytics from Vision API.
+   *
+   * The image is represented as the byte data in the rawImageData byte array, and the annotated
+   * analytics are represented as labelAnnotations, a list of Vision EntityAnnotation objects, which
+   * each represent individual labels.
+   *
+   * Abstraction Function(rawImageData, labelAnnotations) =
+   * An image represented by the bytes in rawImageData, with a list of labels annotated
+   * to the image, represented by the EntityAnnotation(s) in labelAnnotations.
+   *
+   * Representation Invariant:
+   * rawImageData will always represent the same image, it is immutable. It must be non-null. It
+   * must be non-empty. Once set, labelAnnotations will never change, it is immutable. It must be
+   * non-null.
+   *
+   * Safety from Representation Exposure:
+   * All representation fields are immutable.
+   */
+
+  /**
+   * Asserts that the representation invariants are met
+   * The representation invariants are the conditions which must always
+   * be true about the class representation fields.
+   */
+  private void checkRepresentationInvariantMet() {
+    assert rawImageData != null : "raw image data field is null";
+    assert labelAnnotations != null : "label annotations data field is null";
+    assert rawImageData.length != 0 : "raw image data field is empty";
+  }
+
+  /**
+   * Instantiates the VisionManager object parameterized with a byte array of raw image data.
+   *
+   * @param rawImageData The raw image data for the image being represented in this VisionManager
+   *     object. Must be non-empty and non-null.
+   */
+  public VisionManagerImpl(byte[] rawImageData) throws IllegalArgumentException, IOException {
+    this.rawImageData = rawImageData;
+
+    if (rawImageData != null && rawImageData.length != 0) {
+      this.labelAnnotations = detectLabelsFromImageBytes(rawImageData);
```

Comment:
I'd recommend not triggering the call to `detectLabels` from the constructor.

Triggering it independently of construction increases the clarity and flexibility of the class:

1. object construction is expected to be fast, and not involve long-running processes,
2. It would be easier to parallelize calls (currently, with the network call being made serially in the constructor, it could be awkward to try to parallelize multiple vision requests),
3. it's possible that this class will adopt other responsibilities, or allow configuration, that can be or need to be done before the network call,
4. in tests, decoupling processing from construction, you would be able to set up the manager in a `setUp()` method.